sycl/source/detail/device_image_impl.hpp

```diff
@@ -161,12 +161,29 @@ class device_image_impl {
 
   const context &get_context() const noexcept { return MContext; }
 
+  void set_kernel_ids(std::vector<kernel_id> KernelIDs) noexcept {
+    MKernelIDs = std::move(KernelIDs);
+  }
+
   std::vector<kernel_id> &get_kernel_ids_ref() noexcept { return MKernelIDs; }
 
   std::vector<unsigned char> &get_spec_const_blob_ref() noexcept {
     return MSpecConstsBlob;
   }
 
+  RT::PiMem &get_spec_const_buffer_ref() noexcept {
+    std::lock_guard<std::mutex> Lock{MSpecConstAccessMtx};
+    if (nullptr == MSpecConstsBuffer) {
+      const detail::plugin &Plugin = getSyclObjImpl(MContext)->getPlugin();
+      Plugin.call<PiApiKind::piMemBufferCreate>(
+          detail::getSyclObjImpl(MContext)->getHandleRef(),
+          PI_MEM_FLAGS_ACCESS_RW | PI_MEM_FLAGS_HOST_PTR_USE,
+          MSpecConstsBlob.size(), MSpecConstsBlob.data(), &MSpecConstsBuffer,
```

Comment:
This isn't going to work. MSpecConstsBlob doesn't have the default values (and has no way to get them), and the spec constants are not ordered by spec constant ID.

Reply:
Also tagging @AlexeySachkov

Reply:
Yes, per our interface between the runtime and the compiler, DPC++ Tools will use offsets calculated as if spec constants were sorted by their SPIR-V IDs in ascending order, while this `MSpecConstBlob` is built over list of spec constants in device image properties and their order it not defined by our design.

However, technically we can put this to the design doc and update the tools a bit to order properties correctly (we need to use `MapVector` instead of just `Map`)

Reply:
> MSpecConstsBlob doesn't have the default values (and has no way to get them)

Test uses `handler::set_specialization_constant` to write value to `MSpecConstsBlob`. There is not support yet for default values.

> and the spec constants are not ordered by spec constant ID

Tests work locally, possibly it was a lucky coincidence. https://github.com/intel/llvm/pull/3604 should fix this, right?

Reply:
> Tests work locally, possibly it was a lucky coincidence. #3604 should fix this, right?

It shouldn't, that blob was never designed to act like a buffer source.

Reply:
It acts now as a buffer source, there is no problem, but yes, data in this blob wasn't sorted.

> However, technically we can put this to the design doc and update the tools a bit to order properties correctly (we need to use MapVector instead of just Map)

@AlexeySachkov, in this case, can you please add sorting to compile-time, please?

Reply:
@dm-vodopyanov, I've incorporated this into my PR in 92649af4300a4c18f61c4f700a2cf2d981cb64b4, could you please test it with your changes?

Reply:
Thanks @AlexeySachkov, I checked with #3561, still have such error due to probably another issue in the compiler, could be related to wrong order of SPIR-V IDs. This bug can be fixed earlier.